CMakeLists.txt

```diff
@@ -93,6 +93,9 @@ endif(USE_SWIG)
 SET(EIGEN_DIR "${PROJECT_SOURCE_DIR}/eigen")
 include_directories(${EIGEN_DIR})
 
+# See https://gitlab.com/libeigen/eigen/-/blob/master/COPYING.README
+ADD_DEFINITIONS(-DEIGEN_MPL2_ONLY)
```

Comment:
Please add the same eigen define for VS solution file. For example,
https://github.com/microsoft/LightGBM/blob/e95468cbdc5d0732b33f5ffcd61124cb2965d129/windows/LightGBM.vcxproj#L104

Reply:
ok I tried in https://github.com/microsoft/LightGBM/pull/3685/commits/cc06c20ebe0d26fb374eadab44e3b3e49376847c, but I'm not that familiar with this file so please let me know if it's wrong

Reply:
Me neither... 😞 

But I believe we can use global property for `PreprocessorDefinitions` and then in particular configurations append to global value.
https://docs.microsoft.com/en-us/cpp/build/working-with-project-properties?view=msvc-160#predefined-macros
https://github.com/microsoft/LightGBM/blob/e95468cbdc5d0732b33f5ffcd61124cb2965d129/windows/LightGBM.vcxproj#L172

I will try it tomorrow and let you know about results.

Reply:
@jameslamb OK, done in c05f391fd56ab7e703d3226bd71a5bbd4b0c6df6.

Reply:
thanks!

What is the effect of the changes in `windows/LightGBM.vcxproj.filters` from that commit? Should those changes have been made in #3299 ? And if so, is there some way we can catch that in CI?

Reply:
> Should those changes have been made in #3299 ?

Yes.

> And if so, is there some way we can catch that in CI?

No any I'm aware of.

I believe this is more aesthetic fix. I noticed this bug when opened solution file and saw these 2 new files listed outside of the project sources tree. However,  seems that `master` is OK according to the following log lines at Appveyor:
```
running install
INFO:LightGBM:Starting to compile the library.
INFO:LightGBM:Starting to compile with MSBuild from existing solution file.
running build
```

Reply:
hmmm ok.

It seems that maybe this is just a file for configuring an editor and doesn't actually affect how the library is compiled:

* https://docs.microsoft.com/en-us/cpp/build/reference/vcxproj-filters-files?view=msvc-160
* https://stackoverflow.com/a/2841976/3986677

So I guess it makes sense that the build still succeeded without this change. I think maybe we could do something in CI that lists all files in `src/` and checks that they show up in `windows/vcxproj.filters`, but I'd consider that low-priority now that I know it's just a config for an editor.

Reply:
@jameslamb Thanks for the info! I didn't know that
> ... the idea was that we did not want to trigger a rebuild simply because the filters changed, as they don't affect the build.